File: packages/backend/src/check/check.controller.ts
import { Body, Controller, Post } from '@nestjs/common';
import { CheckRequestDto } from './dto/check.dto';
import { CheckService } from './check.service';
import { CitizenService } from 'src/citizen/citizen.service';
import { RuleArgs } from 'src/interfaces/rule';

@Controller('check')
export class CheckController {
  constructor(
    private readonly checkService: CheckService,
    private readonly citizenService: CitizenService,
  ) {}
  @Post()
  async check(@Body() body: CheckRequestDto) {
    const citizen = await this.citizenService.getCitizen(body.citizenID);
    const targetCitizen = body.targetCitizenID
      ? await this.citizenService.getCitizen(body.targetCitizenID)
      : undefined;
    const mentionedCitizens = await Promise.all(
      body.mentionedIDs.map(
        this.citizenService.getCitizen.bind(this.citizenService),
      ),
    );

    const ruleArgs: RuleArgs = {
      citizen,
      message: body.message,
      mentionedCitizens: mentionedCitizens,
      targetCitizen: targetCitizen,
    };

    const { messages, totalSocialChange } = await this.checkService.check(
      ruleArgs,
    );

    this.citizenService.updateSocialCreditScore(
      citizen.discordID,
      totalSocialChange,
    );

    return {
      messages,
    };
  }
}

File: packages/xi-jinping-backend/Controllers/CheckController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using xi_jinping_backend.Dto;
using xi_jinping_backend.Models;
using xi_jinping_backend.Services;
using xi_jinping_backend.Utils;

namespace xi_jinping_backend.Controllers;

[ApiController]
public class CheckController(CitizenContext context, CheckService checkService) : ControllerBase
{
    private readonly CitizenContext _context = context;

    [HttpPost("check")]
    public async Task<CheckResponseDto> Check([FromBody] CheckRequestDto request)
    {
        CitizenItem citizen = await _context.FindOrCreate(request.CitizenID, request.CitizenUsername);
        await citizen.AddGuild(request.Guild.Id, _context);

        CitizenItem? targetCitizen = request.TargetCitizenID != null
            ? await _context.CitizenItems.FirstOrDefaultAsync(x => x.DiscordId == request.TargetCitizenID)
            : null;

        CitizenItem[] mentionedCitizens = [];
        foreach (string id in request.MentionedIDs)
        {
            CitizenItem? mentionedCitizen = await _context.CitizenItems.FirstOrDefaultAsync(x => x.DiscordId == id);
            if (mentionedCitizen == null) continue;

            await mentionedCitizen.AddGuild(request.Guild.Id, _context);
            mentionedCitizens = [.. mentionedCitizens, mentionedCitizen];
        }

        RuleArgs ruleArgs = new(citizen, targetCitizen, mentionedCitizens, request.Guild.Id);

        var (messages, change) = await checkService.Check(ruleArgs);

        if (change != 0)
        {
            citizen.SocialCreditScore += change;
            await _context.SaveChangesAsync();
        }

        return new CheckResponseDto
        {
            Messages = messages
        };
    }
}
